File: hdf-converters/registry/registry_test.go
package registry

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func makeFP(id string) ConverterFingerprint {
	return ConverterFingerprint{
		ID:          id,
		Label:       "Test",
		Direction:   DirectionIngest,
		InputFamily: FamilyJSON,
		OutputType:  OutputResults,
		Fingerprint: func(input any) float64 { return 1.0 },
	}
}

func TestRegister(t *testing.T) {
	t.Run("registers a fingerprint", func(t *testing.T) {
		ResetRegistry()
		Register(makeFP("test-a"))
		assert.Len(t, GetFingerprints(), 1)
	})

	t.Run("registers multiple fingerprints", func(t *testing.T) {
		ResetRegistry()
		Register(makeFP("a"))
		Register(makeFP("b"))
		assert.Len(t, GetFingerprints(), 2)
	})

	t.Run("panics on duplicate ID", func(t *testing.T) {
		ResetRegistry()
		Register(makeFP("dup"))
		assert.PanicsWithValue(t, "duplicate fingerprint: dup", func() {
			Register(makeFP("dup"))
		})
	})
}

func TestGetFingerprints(t *testing.T) {
	t.Run("returns empty slice when none registered", func(t *testing.T) {
		ResetRegistry()
		assert.Empty(t, GetFingerprints())
	})

	t.Run("returns all registered", func(t *testing.T) {
		ResetRegistry()
		Register(makeFP("a"))
		Register(makeFP("b"))
		assert.Len(t, GetFingerprints(), 2)
	})
}

func TestGetIngestFingerprints(t *testing.T) {
	ResetRegistry()
	Register(makeFP("in"))
	fp := makeFP("out")
	fp.Direction = DirectionExport
	Register(fp)

	ingest := GetIngestFingerprints()
	assert.Len(t, ingest, 1)
	assert.Equal(t, "in", ingest[0].ID)
}

func TestGetFingerprint(t *testing.T) {
	t.Run("finds by ID", func(t *testing.T) {
		ResetRegistry()
		Register(makeFP("sarif-to-hdf"))
		fp := GetFingerprint("sarif-to-hdf")
		require.NotNil(t, fp)
		assert.Equal(t, "Test", fp.Label)
	})

	t.Run("returns nil for unknown ID", func(t *testing.T) {
		ResetRegistry()
		assert.Nil(t, GetFingerprint("nonexistent"))
	})
}

func TestOutputType(t *testing.T) {
	ResetRegistry()
	fp1 := makeFP("oscal-sar")
	fp1.OutputType = OutputResults
	fp2 := makeFP("oscal-poam")
	fp2.OutputType = OutputAmendments
	Register(fp1)
	Register(fp2)

	assert.Equal(t, OutputResults, GetFingerprint("oscal-sar").OutputType)
	assert.Equal(t, OutputAmendments, GetFingerprint("oscal-poam").OutputType)
}

func TestResetRegistry(t *testing.T) {
	ResetRegistry()
	Register(makeFP("a"))
	Register(makeFP("b"))
	assert.Len(t, GetFingerprints(), 2)
	ResetRegistry()
	assert.Empty(t, GetFingerprints())
}

File: hdf-converters/shared/typescript/registry.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import {
  registerFingerprint,
  getFingerprints,
  getIngestFingerprints,
  getFingerprint,
  _resetRegistry,
  type ConverterFingerprint,
} from './registry.js';

function makeFP(overrides: Partial<ConverterFingerprint> = {}): ConverterFingerprint {
  return {
    id: 'test-converter',
    label: 'Test',
    direction: 'ingest',
    inputFamily: 'json',
    outputType: 'results',
    fingerprint: () => 1.0,
    ...overrides,
  };
}

describe('registry', () => {
  beforeEach(() => {
    _resetRegistry();
  });

  describe('registerFingerprint', () => {
    it('registers a fingerprint', () => {
      registerFingerprint(makeFP());
      expect(getFingerprints()).toHaveLength(1);
    });

    it('registers multiple fingerprints', () => {
      registerFingerprint(makeFP({ id: 'a' }));
      registerFingerprint(makeFP({ id: 'b' }));
      expect(getFingerprints()).toHaveLength(2);
    });

    it('throws on duplicate ID', () => {
      registerFingerprint(makeFP({ id: 'dup' }));
      expect(() => registerFingerprint(makeFP({ id: 'dup' }))).toThrow(
        'Duplicate fingerprint: dup'
      );
    });
  });

  describe('getFingerprints', () => {
    it('returns empty array when no fingerprints registered', () => {
      expect(getFingerprints()).toHaveLength(0);
    });

    it('returns all registered fingerprints', () => {
      registerFingerprint(makeFP({ id: 'a', direction: 'ingest' }));
      registerFingerprint(makeFP({ id: 'b', direction: 'export' }));
      expect(getFingerprints()).toHaveLength(2);
    });
  });

  describe('getIngestFingerprints', () => {
    it('filters to ingest-only', () => {
      registerFingerprint(makeFP({ id: 'in', direction: 'ingest' }));
      registerFingerprint(makeFP({ id: 'out', direction: 'export' }));
      const ingest = getIngestFingerprints();
      expect(ingest).toHaveLength(1);
      expect(ingest[0].id).toBe('in');
    });
  });

  describe('getFingerprint', () => {
    it('returns fingerprint by ID', () => {
      registerFingerprint(makeFP({ id: 'sarif-to-hdf', label: 'SARIF' }));
      const result = getFingerprint('sarif-to-hdf');
      expect(result).toBeDefined();
      expect(result!.label).toBe('SARIF');
    });

    it('returns undefined for unknown ID', () => {
      expect(getFingerprint('nonexistent')).toBeUndefined();
    });
  });

  describe('outputType', () => {
    it('stores outputType on registered fingerprint', () => {
      registerFingerprint(makeFP({ id: 'oscal-sar', outputType: 'results' }));
      registerFingerprint(makeFP({ id: 'oscal-poam', outputType: 'amendments' }));
      expect(getFingerprint('oscal-sar')!.outputType).toBe('results');
      expect(getFingerprint('oscal-poam')!.outputType).toBe('amendments');
    });
  });

  describe('_resetRegistry', () => {
    it('clears all registered fingerprints', () => {
      registerFingerprint(makeFP({ id: 'a' }));
      registerFingerprint(makeFP({ id: 'b' }));
      expect(getFingerprints()).toHaveLength(2);
      _resetRegistry();
      expect(getFingerprints()).toHaveLength(0);
    });
  });
});
